discord/emoji.py

```diff
@@ -65,7 +66,14 @@ class PartialEmoji(namedtuple('PartialEmoji', 'animated name id')):
         The ID of the custom emoji, if applicable.
     """
 
-    __slots__ = ()
+    __slots__ = ('animated', 'name', 'id')
+
+    def __init__(self, state, *, animated, name, id=None):
```

Comment:
It seems to me that people do not want this to be changed since they use it for their own purposes. I don't consider the use case invalid but it does put a bit of a dent in how you could implement this so it works fine for both groups of people.

After some thinking I've come up with the following solution:

1. Don't inherit from `Asset`.
2. Make `PartialEmoji.url` return an `Asset` instead.
3. Make a classmethod that constructs the `PartialEmoji` for the library's internal uses with the `ConnectionState` being passed in.
4. Default to `None` for the `PartialEmoji._state`
5. When calling `PartialEmoji.url`, the `None` connection state should propagate over to it.
6. Inside `Asset.save` check if the state parameter is `None` and raise if so.

This should satisfy both parties and make the behaviour consistent (it's an asset) but also raises during invalid usage.

As another note, since `namedtuple` was removed from the inheritance chain, this class no longer provides a `__ne__` as per the documentation. (Also it seems to me that `Emoji.__ne__` does not exist either despite being documented so maybe get that one as well)

Reply:
If not provided, `__ne__` defaults to the negation of `__eq__` so that's not necessary.

Reply:
I was nearly done with it and was about to test everything, but then I realised it would cause an inconsistency between `Emoji` being an `Asset` itself, and `PartialEmoji.url` returning an `Asset`. This could affect people who do `typing.Union[discord.Emoji, discord.PartialEmoji]` for example.

Should I also revert `Emoji` to match with `PartialEmoji`?

Reply:
Yeah probably. If truly necessary there can be `save` helper functions that just redirect to the appropriate `Asset.save` functions.

Reply:
Gotcha, pre-emptively reverted it anyway. Should be available in the next commit!